sdk/core/System.ClientModel/src/Pipeline/MessageLoggingPolicy.cs

```diff
@@ -0,0 +1,371 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.ClientModel.Internal;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.Tracing;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+
+namespace System.ClientModel.Primitives;
+
+/// <summary>
+/// A <see cref="PipelinePolicy"/> used by a <see cref="ClientPipeline"/> to
+/// log request and response information.
+/// </summary>
+public class MessageLoggingPolicy : PipelinePolicy
+{
+    /// <summary>
+    /// The <see cref="ClientRetryPolicy"/> instance used by a default
+    /// <see cref="ClientPipeline"/>.
+    /// </summary>
+    public static MessageLoggingPolicy Default { get; } = new MessageLoggingPolicy();
+
+    private readonly ClientLoggingOptions _loggingOptions;
+    private PipelineMessageLogger? _messageLogger;
+    private readonly string _clientAssembly = typeof(MessageLoggingPolicy).Assembly.GetName().Name!;
+
+    private bool _enableMessageContentLogging => _loggingOptions.EnableMessageContentLogging ?? ClientLoggingOptions.DefaultEnableMessageContentLogging;
+    private int _maxLength => _loggingOptions.MessageContentSizeLimit ?? ClientLoggingOptions.DefaultMessageContentSizeLimit;
+
+    /// <summary>
+    /// Creates a new instance of the <see cref="MessageLoggingPolicy"/> class.
+    /// </summary>
+    /// <param name="options">The user-provided logging options object.</param>
+    public MessageLoggingPolicy(ClientLoggingOptions? options = default)
+    {
+        _loggingOptions = options ?? new();
+    }
+
+    /// <inheritdoc/>
+    public override void Process(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex) =>
+        ProcessSyncOrAsync(message, pipeline, currentIndex, async: false).EnsureCompleted();
+
+    /// <inheritdoc/>
+    public override async ValueTask ProcessAsync(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex) =>
+        await ProcessSyncOrAsync(message, pipeline, currentIndex, async: true).ConfigureAwait(false);
+
+    private async ValueTask ProcessSyncOrAsync(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex, bool async)
+    {
+        _messageLogger ??= new PipelineMessageLogger(_loggingOptions.GetPipelineMessageSanitizer(), _loggingOptions.LoggerFactory);
+
+        if (!_messageLogger.IsEnabled(LogLevel.Warning, EventLevel.Warning)) // Warning is the highest level logged
```

Comment:
I'm not sure that I understand what's happening here.   We're not chaining the next policy if we don't have warning logging enabled?

A comment may be helpful to explain the flow.

Reply:
I'll add a comment. This checks if at least warnings are enabled. If not, then it immediately skips to the next policy. If warnings or lower are enabled, then process next is called after the request info is logged